src/lib/storage/buffer/volatile_region.hpp

```diff
@@ -0,0 +1,105 @@
+#pragma once
+
+#include <atomic>
+#include <boost/dynamic_bitset.hpp>
+#include <forward_list>
+#include <memory>
+#include <mutex>
+#include <span>
+#include <tuple>
+#include "frame.hpp"
+#include "page_id.hpp"
+#include "types.hpp"
+
+/**
+ * @brief The VolatileRegion class provides methods for page and frame access and NUMA page movement for a defined page size type. 
+ * 
+ * The class uses MADV_FREE_REUSABLE and MADV_DONTNEED on Linux to free pages. On OS X, it uses MADV_FREE_REUSABLE to free pages and MADV_FREE_REUSE to mark pages as reusable. The class provides two methods for freeing pages: free() and reuse(). The free() method uses MADV_FREE_REUSABLE on OS X and MADV_DONTNEED on Linux to free a page. The reuse() method uses MADV_FREE_REUSE on OS X to mark a page as reusable. This method is not implemented on Linux.
+ * 
+ * The class provides three methods for moving pages between NUMA nodes: mbind_to_numa_node(), move_page_to_numa_node(), and memcpy_page_to_numa_node(). The mbind_to_numa_node() method uses mbind to move a page to a NUMA node. This method is not portable and only works on Linux with NUMA support. The move_page_to_numa_node() method uses move_pages to move a page to a NUMA node. This method is not portable and only works on Linux with NUMA support. The memcpy_page_to_numa_node() method uses double memcpy + madvise(MADV_DONTNEED) with an intermediate buffer to move a page to a NUMA node. This method is faster with more concurrency.
+ */
+namespace hyrise {
+class VolatileRegion final : public Noncopyable {
+ public:
+  // Default size of the virtual memory region for all page size types. Default is 256 GB.
+  constexpr static uint64_t DEFAULT_RESERVED_VIRTUAL_MEMORY = 1UL << 38;
+  constexpr static uint64_t DEFAULT_RESERVED_VIRTUAL_MEMORY_PER_REGION =
+      (DEFAULT_RESERVED_VIRTUAL_MEMORY / PAGE_SIZE_TYPES_COUNT) / bytes_for_size_type(MAX_PAGE_SIZE_TYPE) *
+      bytes_for_size_type(MAX_PAGE_SIZE_TYPE);
+
+  using MappedRegion = std::span<std::byte, DEFAULT_RESERVED_VIRTUAL_MEMORY>;
+  using MappedPageRegion = std::span<std::byte, DEFAULT_RESERVED_VIRTUAL_MEMORY_PER_REGION>;
+
+  // Enable mprotect class for debugging purposes
+  constexpr static bool ENABLE_MPROTECT = false;
+
+  // Create a VolatileRegion in a memory region for a given size_type.
+  VolatileRegion(const PageSizeType size_type, MappedPageRegion region);
```

Comment:
Constructing the class and requiring to pass a data structure that is part of the class I am about to construct (MappedPageRegion) is strange.